src/components/App/ImportStudyList.js

```diff
@@ -0,0 +1,168 @@
+import {
+    Button,
+    Dialog,
+    DialogActions,
+    DialogContent,
+    DialogContentText,
+    DialogTitle,
+    TextField,
+} from '@material-ui/core';
+import { queryWebsoc } from '../../helpers';
+import RightPaneStore from '../../stores/RightPaneStore';
+import { addCourse, openSnackbar } from '../../actions/AppStoreActions';
+import AppStore from '../../stores/AppStore';
+import { GetApp } from '@material-ui/icons';
+import { termData } from '../../termData';
+import MenuItem from '@material-ui/core/MenuItem';
+import Select from '@material-ui/core/Select';
+import InputLabel from '@material-ui/core/InputLabel';
+
+class ImportStudyList extends PureComponent {
+    state = {
+        isOpen: false,
+        selectedTerm: RightPaneStore.getFormData().term,
+        studyListText: '',
+    };
+
+    handleChange = (event) => {
+        this.setState({ selectedTerm: event.target.value });
+    };
+
+    handleError = (error) => {
+        openSnackbar('error', 'An error occurred while trying to import the Study List.');
+        console.error(error);
+    };
+
+    handleOpen = () => {
+        this.setState({ isOpen: true });
+    };
+
+    handleClose = async (doImport) => {
+        this.setState({ isOpen: false }, () => {
+            document.removeEventListener('keydown', this.enterEvent, false);
+            this.setState({ studyListText: '' });
+        });
+        if (doImport) {
+            document.removeEventListener('keydown', this.enterEvent, false);
+            if (!this.state.studyListText.match(/\d{5}/g)) {
+                openSnackbar('error', 'Cannot import an empty/invalid Study List.');
+                return;
+            }
+            const currSchedule = AppStore.getCurrentScheduleIndex();
+            const sectionCodes = this.state.studyListText
+                .split('\n')
+                .map((line) => line.match(/\d{5}/g))
+                .filter((id) => id)
+                .flat();
+            let sectionsAdded = 0;
+            try {
+                (
+                    await Promise.all(
+                        sectionCodes
+                            .reduce((result, item, index) => {
+                                // WebSOC queries can have a maximum of 8 course codes in tandem
+                                const chunkIndex = Math.floor(index / 8);
+                                result[chunkIndex] ? result[chunkIndex].push(item) : (result[chunkIndex] = [item]);
+                                return result;
+                            }, []) // https://stackoverflow.com/a/37826698
+                            .map((sectionCode) =>
+                                queryWebsoc({ term: this.state.selectedTerm, sectionCodes: sectionCode.join(',') })
+                            )
+                    )
+                ).forEach((response) => {
+                    response.schools
+                        .map((school) => school.departments)
```

Comment:
Since you aren't returning anything here, using repeated map/flatten is more confusing than just using normal loops. I recommend refactoring it for readability. See this example of going through the websoc data from `helpers.js`:
```
for (const school of jsonResp.schools) {
      for (const department of school.departments) {
          for (const course of department.courses) {
              for (const section of course.sections) {
                  addedCourses.push({
                      ...sectionCodeToInfoMapping[section.sectionCode],
                      deptCode: department.deptCode,
                      courseNumber: course.courseNumber,
                      courseTitle: course.courseTitle,
                      courseComment: course.courseComment,
                      prerequisiteLink: course.prerequisiteLink,
                      section: section,
                  });
              }
          }
      }
  }

```
Also since we're having to do this twice, it would be nice to write a function to encapsulate extracting the individual course info from a websoc query. You can do this yourself or create an issue for it if you want.

Reply:
Will do. Is it best practice to make a separate branch/PR for new functionality separate from this feature (e.g. this and refactoring `TermSelector` like you mentioned [here](https://github.com/icssc-projects/AntAlmanac/pull/302#discussion_r828435583)), or is it fine to just commit to this branch?

Reply:
Normally I just add it to the same branch. Not sure if that is best practice though. @ChaseC99 what are your thoughts on when to make an entirely new branch vs just lumping a change in with an existing branch that isn't entirely related?

Reply:
I usually try to keep my PRs as small as possible, so it's easier for others to review.

In this case, let's create a new issue, leave a todo comment in the code, and immediately open a new PR after merging this one.

That helper function will affect core functionality, so it'd probably be best to test it thoroughly in a separate PR.

Reply:
Done. I've also created issues #308 and #309 for this and the TermSelector issue Eric mentioned below.